### PR TITLE
V14

### PR DESCRIPTION
**Pull request by brynrhodes**

### CODE REVIEW

**Comment by cmoesel** [comment]
This is a pretty large PR.  I'm just coming off a few days of vacation and July 4 is around the corner.  It looks like I'm the lone reviewer on it -- when do you need it and is there anyone else who might be able to help?

**Comment by brynrhodes** [comment]
@cmoesel , yes, sorry, this is the release merge to master, so the code has all passed unit testing, full regression, as well as integration testing with the targets. It's more a quality gate to make sure we're ready and agree to merge the 1.4 branch into master than a full code review. Does that help?

**Comment by cmoesel** [comment]
OK.  Having the understanding that this has already been reviewed in some form, and knowing that Bonnie has also been testing these releases along the way, I'll plan to do a light review today -- and if there are no glaring issues, I'll approve.

File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/LibraryBuilder.java [hunk 1]
```diff
@@ -2114,18 +2201,18 @@ private DataType getExpressionDefResultType(ExpressionDef expressionDef) {
             return expressionDef.getResultType();
         }
 
-        // If the current expression context is patient, a reference to a population context expression will indicate a full
+        // If the current expression context is specific, a reference to an unspecified context expression will indicate a full
```
Comment (cmoesel) [inline Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/LibraryBuilder.java]: Only a comment, but probably should replace `unspecified` with `unfiltered`

File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/LibraryBuilder.java [hunk 2]
```diff
@@ -2114,18 +2201,18 @@ private DataType getExpressionDefResultType(ExpressionDef expressionDef) {
...
         // evaluation of the population context expression, and the result type is the same.
-        if (inPatientContext()) {
+        if (inSpecificContext()) {
             return expressionDef.getResultType();
         }
 
-        // If the current expression context is population, a reference to a patient context expression will need to be
-        // performed for every patient in the population, so the result type is promoted to a list (if it is not already).
-        if (inPopulationContext()) {
+        // If the current expression context is unspecified, a reference to a specific context expression will need to be
```
Comment (cmoesel) [inline Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/LibraryBuilder.java]: Only a comment, but probably should replace `unspecified` with `unfiltered`

File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/preprocessor/CqlPreprocessorVisitor.java [hunk 3]
```diff
@@ -96,13 +105,21 @@ public Object visitParameterDefinition(@NotNull cqlParser.ParameterDefinitionCon
...
+        String modelIdentifier = ctx.modelIdentifier() != null ? (String)visit(ctx.modelIdentifier()) : null;
+        String unqualifiedContext = (String)visit(ctx.identifier());
+        if (modelIdentifier != null && !modelIdentifier.equals("")) {
+            currentContext = modelIdentifier + "." + unqualifiedContext;
+        }
+        else {
+            currentContext = unqualifiedContext;
+        }
+
+        if (!implicitContextCreated && !unqualifiedContext.equals("Unspecified")) {
```
Comment (cmoesel) [inline Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/preprocessor/CqlPreprocessorVisitor.java]: I think this should be `"Unfiltered"`

File: Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/OperatorTests/CqlComparisonOperators.cql [hunk 4]
```diff
@@ -1,6 +1,6 @@
 library ComparisonOperator version '1'
 
-context Population
+context Unspecified
```
Comment (cmoesel) [inline Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/OperatorTests/CqlComparisonOperators.cql]: I think this should be `Unfiltered`

File: Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/OperatorTests/CqlIntervalOperators.cql [hunk 5]
```diff
@@ -1,6 +1,6 @@
 library CqlIntervalOperatorsTest version '1'
 
-context Population
+context Unspecified
```
Comment (cmoesel) [inline Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/OperatorTests/CqlIntervalOperators.cql]: I think this should be `Unfiltered`

File: Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/OperatorTests/CqlListOperators.cql [hunk 6]
```diff
@@ -1,6 +1,6 @@
 library ListOperator version '1'
 
-context Population
+context Unspecified
```
Comment (cmoesel) [inline Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/OperatorTests/CqlListOperators.cql]: I think this should be `Unfiltered`

File: Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/TestPatientContext.cql [hunk 7]
```diff
@@ -6,6 +6,6 @@ context Patient
 
 define Patient: true
 
-context Population
+context Unspecified
```
Comment (cmoesel) [inline Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/TestPatientContext.cql]: I think this should be `Unfiltered`

**Comment by cmoesel** [comment]
FYI -- I think the Bonnie team ran into some issues related to `Unfiltered` vs `Unspecified` when updating our CQL engine test cases.  Converting QUICK-based CQL to use `Unfiltered` context instead of `Population` context caused an error in the CQL-to-ELM, while using `Unspecified` _worked_.  I expect it may be related to the issues I pointed out above.

File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/preprocessor/CqlPreprocessorVisitor.java [hunk 8]
```diff
@@ -6,18 +6,22 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+
 public class CqlPreprocessorVisitor extends cqlBaseVisitor {
     private LibraryInfo libraryInfo = new LibraryInfo();
-    private boolean implicitPatientCreated = false;
-    private String currentContext = "Patient";
+    private boolean implicitContextCreated = false;
+    private String currentContext = "Unspecified";
```
[the remark this answers is not in the document]
Comment (brynrhodes) [reply Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/preprocessor/CqlPreprocessorVisitor.java]: Agreed, corrected

File: Src/java/tools/cql-formatter/src/test/resources/org/cqframework/cql/tools/formatter/comment-after.cql [hunk 9]
```diff
@@ -1,7 +1,7 @@
 define x: 'Shiver my timbers, shiver my soul'
 // comment
 
-context Population
+context Unspecified
```
[the remark this answers is not in the document]
Comment (brynrhodes) [reply Src/java/tools/cql-formatter/src/test/resources/org/cqframework/cql/tools/formatter/comment-after.cql]: Agreed, corrected

File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/Cql2ElmVisitor.java [hunk 10]
```diff
@@ -658,52 +671,53 @@ public ConceptDef visitConceptDefinition(@NotNull cqlParser.ConceptDefinitionCon
...
-                implicitPatientExpressionDef.getExpression().setResultType(libraryBuilder.resolveTypeName("System", "Any"));
-                implicitPatientExpressionDef.setResultType(implicitPatientExpressionDef.getExpression().getResultType());
-                libraryBuilder.addExpression(implicitPatientExpressionDef);
-            }
+        String modelIdentifier = parseString(ctx.modelIdentifier());
+        String unqualifiedIdentifier = parseString(ctx.identifier());
+
+        currentContext = modelIdentifier != null ? modelIdentifier + "." + unqualifiedIdentifier : unqualifiedIdentifier;
+
+        if (!unqualifiedIdentifier.equals("Unspecified")) {
```
[the remark this answers is not in the document]
Comment (brynrhodes) [reply Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/Cql2ElmVisitor.java]: Agreed, corrected

File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/LibraryBuilder.java (hunk 1 above)
Comment (brynrhodes) [reply Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/LibraryBuilder.java]: Agreed, corrected

File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/LibraryBuilder.java (hunk 2 above)
Comment (brynrhodes) [reply Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/LibraryBuilder.java]: Agreed, corrected

File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/preprocessor/CqlPreprocessorVisitor.java (hunk 3 above)
Comment (brynrhodes) [reply Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/preprocessor/CqlPreprocessorVisitor.java]: Agreed, corrected

File: Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/OperatorTests/CqlComparisonOperators.cql (hunk 4 above)
Comment (brynrhodes) [reply Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/OperatorTests/CqlComparisonOperators.cql]: Agreed, corrected

File: Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/OperatorTests/CqlIntervalOperators.cql (hunk 5 above)
Comment (brynrhodes) [reply Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/OperatorTests/CqlIntervalOperators.cql]: Agreed, corrected

File: Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/OperatorTests/CqlListOperators.cql (hunk 6 above)
Comment (brynrhodes) [reply Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/OperatorTests/CqlListOperators.cql]: Agreed, corrected

File: Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/TestPatientContext.cql (hunk 7 above)
Comment (brynrhodes) [reply Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/TestPatientContext.cql]: Agreed, corrected

**Comment by brynrhodes** [comment]
@cmoesel , yes, the Bonnie team reached out and confirmed they have encountered this issue. I've done a global search for Unspecified to make sure I didn't miss any more. Thank you for the review!

**Comment by cmoesel** [comment]
@jbradl11 -- have you had an opportunity to test this updated code and confirm that it resolves the issues you were encountering?  Looks good from my review, but wanted to see if you had tried it.

**Comment by jbradl11** [comment]
@cmoesel I haven't seen an updated JAR since the issue was discovered.

**Comment by brynrhodes** [comment]
@jbradl11 , @cmoesel , just released 1.4.7-SNAPSHOT with the updates from this review.

**Comment by brynrhodes** [comment]
@cmoesel , @jbradl11 , are we okay to merge this PR?

**Comment by cmoesel** [comment]
It's fine by me!

**Comment by jbradl11** [comment]
That works for me as well. I am going to tag someone from Semantic Bits for awareness (the Bonnie contractor). @AndrewBird81

**Comment by brynrhodes** [comment]
@cmoesel , can I trouble you to mark the PR approved?

**Pull request merged**